File: DSA/DrishtiSeth_590011761/Day_31/Problem2.py
def nextGreaterElement(arr):
    n = len(arr)
    ans = [-1] * n
    stack = []

    for i in range(n - 1, -1, -1):
        while stack and stack[-1] <= arr[i]:
            stack.pop()

        if stack:
            ans[i] = stack[-1]

        stack.append(arr[i])

    return ans


# Example
arr = [1, 3, 2, 4]
print(nextGreaterElement(arr))

File: DSA/MrigankSingh_590012123/Day31/Problem2.java
import java.util.Stack;

public class Problem2 {
    public int[] nextGreaterElements(int[] arr) {
        int[] answer = new int[arr.length];
        Stack<Integer> stack = new Stack<>();

        for (int i = arr.length - 1; i >= 0; i--) {
            while (!stack.isEmpty() && stack.peek() <= arr[i]) {
                stack.pop();
            }

            answer[i] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(arr[i]);
        }

        return answer;
    }
}